mungegithub/mungers/publish_scripts/util.sh

```diff
@@ -19,110 +19,446 @@
 set -o errexit
 set -o nounset
 set -o pipefail
+set -o xtrace
 
-# sync_repo() cherry picks the latest changes in k8s.io/kubernetes/${filter} to the
+# sync_repo() cherry picks the latest changes in k8s.io/kubernetes/<repo> to the
 # local copy of the repository to be published.
 #
 # prerequisites
 # 1. we are in the root of the repository to be published
-# 2. we are on the branch to be published (let's call it "target-branch")
+# 2. we are on the branch to be published (let's call it "target-branch"), possibly this
+#    is an orphaned branch for new branches.
 # overall flow
-# 1. fetch the current level of k8s.io/kubernetes
-# 2. check out the $src_branch of k8s.io/kubernetes as branch kube-sync
-# 3. rewrite the history of branch kube-sync to *only* include code in $subdirectory
-# 4. locate all commits between the last time we sync'ed and now
-# 5. switch back to the "target-branch"
-# 6. for each commit, cherry-pick it (which will keep authorship) into "target-branch"
-# 7. update metadata files indicating which commits we've sync'ed to
+# 1.    fetch the current level of k8s.io/kubernetes
+# 2.    check out the $src_branch of k8s.io/kubernetes as branch filter-branch
+# 3.    rewrite the history of branch filter-branch to *only* include code in $subdirectory
+# 4.    locate all commits between the last time we sync'ed and now **on the mainline**, i.e.
+#       using first-parent ancestors only, not going into a PR branch.
+# 5.    switch back to the "target-branch"
+# 6a.   single commits which are not merges are just applied. If they have had a merge above to
+#       be merged into the mainline, those merge commits (which might be dropped by the filter step)
```

Comment:
Yes. The merge commit exists in kube, but is an empty ff merge. Git filter-branch will drop it. But it is import for the published repo as it can be the target of a tag or fork of a release-branch later.